src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.GateThread.cs

```diff
@@ -0,0 +1,128 @@
+
+                    ThreadPoolInstance._cpuUtilization = cpu.CurrentUtilization;
+
+                    if (!s_disableStarvationDetection)
+                    {
+                        if (ThreadPoolInstance._numRequestedWorkers > 0 && SufficientDelaySinceLastDequeue())
+                        {
+                            ThreadCounts counts = ThreadCounts.VolatileReadCounts(ref ThreadPoolInstance._separated.counts);
+                            // don't add a thread if we're at max or if we are already in the process of adding threads
+                            while (counts.numExistingThreads < ThreadPoolInstance._maxThreads && counts.numExistingThreads >= counts.numThreadsGoal)
+                            {
+                                if (s_debuggerBreakOnWorkStarvation)
+                                {
+                                    Debug.WriteLine("The CLR ThreadPool detected work starvation!");
+                                    Debugger.Break();
+                                }
+
+                                ThreadCounts newCounts = counts;
+                                newCounts.numThreadsGoal = (short)(newCounts.numExistingThreads + 1);
+                                ThreadCounts oldCounts = ThreadCounts.CompareExchangeCounts(ref ThreadPoolInstance._separated.counts, newCounts, counts);
+                                if (oldCounts == counts)
+                                {
+                                    HillClimbing.ThreadPoolHillClimber.ForceChange(newCounts.numThreadsGoal, HillClimbing.StateOrTransition.Starvation);
+                                    WorkerThread.MaybeAddWorkingWorker();
+                                    break;
+                                }
+                                counts = oldCounts;
+                            }
+                        }
+                    }
+                }
+            }
+
+            // called by logic to spawn new worker threads, return true if it's been too long
+            // since the last dequeue operation - takes number of worker threads into account
+            // in deciding "too long"
+            private static bool SufficientDelaySinceLastDequeue()
+            {
+                int delay = Environment.TickCount - Volatile.Read(ref ThreadPoolInstance._separated.lastDequeueTime);
+
+                int minimumDelay;
+
+                if(ThreadPoolInstance._cpuUtilization < CpuUtilizationLow)
+                {
+                    minimumDelay = GateThreadDelayMs;
+                }
+                else
+                {
+                    ThreadCounts counts = ThreadCounts.VolatileReadCounts(ref ThreadPoolInstance._separated.counts);
+                    int numThreads = counts.numThreadsGoal;
+                    minimumDelay = numThreads * DequeueDelayThreshold;
+                }
+
+                return delay > minimumDelay;
+            }
+
+            private static void CreateGateThread()
```

Comment:
Or separate the part the actually creates the thread into this method and put the rest in EnsureRunning